src/main/java/com/commercetools/sync/producttypes/ProductTypeSync.java

```diff
@@ -103,57 +125,55 @@ public ProductTypeSync(@Nonnull final ProductTypeSyncOptions productTypeSyncOpti
     @Override
     protected CompletionStage<ProductTypeSyncStatistics> processBatch(@Nonnull final List<ProductTypeDraft> batch) {
 
-        final Set<ProductTypeDraft> validProductTypeDrafts = batch.stream()
-                                                                  .filter(this::validateDraft)
-                                                                  .collect(toSet());
+        readyToResolve = ConcurrentHashMap.newKeySet();
+        final ProductTypeBatchProcessor batchProcessor = new ProductTypeBatchProcessor(batch, this);
+        batchProcessor.validateBatch();
 
-        if (validProductTypeDrafts.isEmpty()) {
-            statistics.incrementProcessed(batch.size());
-            return completedFuture(statistics);
-        } else {
-            final Set<String> keys = validProductTypeDrafts.stream().map(ProductTypeDraft::getKey).collect(toSet());
+        final Set<String> keysToCache = batchProcessor.getKeysToCache();
 
+        return productTypeService
+            .cacheKeysToIds(keysToCache)
+            .handle(ImmutablePair::new)
```

Comment:
because I want to do `thenCompose` to execute another completionstage as soon as handle results. And using ImmutablePair is convenient for me to get the 2 params (exception and result) in one object that I can use in a `thenCompose`